### DISCUSSION TITLE
sabbrev expansion with <enter>

### DISCUSSION THREAD

**Discussion by pl643** [akinomyoga/ble.sh#138] [General]
Just discover ble.sh and wondered if it would possible have sabbrev expand with the <Enter> key also.

The current behavior is good with the <Space>

```
$ ble-sabbrev l='ls -l'
$ l<Space> - expand to 'ls -l '     * good behavior *
$ l<Enter>
l: command not found
[ble: exit 127]
```

I like to have the below rehavior also:

```
$ l<Enter>
$ ls -l
-rw-r--r-- 1 ply ply  51966 Sep 16 16:08 ble.pp
-rw-r--r-- 1 ply ply  29823 Sep 16 16:08 blerc
.
.
```

Thank you.

**Comment by akinomyoga** [comment]
Thank you for the feedback!

**Q1**: May I ask which behavior do you precisely intend?

1. When there are some words which can be expanded by `sabbrev`, <kbd>RET</kbd> will expand the words, and the user can continue to edit the command line. Finally, when there are no expandable words by `sabbrev` when the user hits <kbd>RET</kbd>, the command will be executed.

```bash
$ ble-sabbrev l='ls -l'[RET]
$ l[RET]
$ ls -l[RET]
-rw-r--r-- 1 ply ply 51966 Sep 16 16:08 ble.pp
-rw-r--r-- 1 ply ply 29823 Sep 16 16:08 blerc
...
```

2. When there are words expandable by `sabbrev`, <kbd>RET</kbd> will show the expanded result of the command line and executes the expanded command immediately just like history expansions.

```bash
$ ble-sabbrev l='ls -l'[RET]
$ l[RET]
[ble: expand] ls -l
-rw-r--r-- 1 ply ply 51966 Sep 16 16:08 ble.pp
-rw-r--r-- 1 ply ply 29823 Sep 16 16:08 blerc
...
```

3. When there are words expandable by `sabbrev`, <kbd>RET</kbd>internally expands the command line and executes the command immediately without showing the expanded result just like alias expansions (I haven't tried but `zsh-abbr` seems to behave like this according to its README).

```bash
$ ble-sabbrev l='ls -l'[RET]
$ l[RET]
-rw-r--r-- 1 ply ply 51966 Sep 16 16:08 ble.pp
-rw-r--r-- 1 ply ply 29823 Sep 16 16:08 blerc
...
```

**Q2**: Should it happen also for the words that are not in the position of commands?

```bash
$ ble-sabbrev '\X'='hello world'[RET]
$ echo \X[RET]
```

**Q3**: Should the cursor position matter for whether the sabbrev expansion happens? I.e., should it happen also when the cursor isn't in the front of the word to be expanded?

```bash
# Example 1

$ ble-sabbrev l='ls -l'[RET]
$ l file1[RET]

# Example 2 (going to the beginning of the command line before pressing RET)

$ ble-sabbrev l='ls -l'[RET]
$ [RET]l
```

It seemed like useful in the first thought, and also it's technically possible. But this might not work well because unexpected expansions may happen by the behavior. Also, there is no way to literally run a command that contains words registered in `sabbrev`.

**Comment by bkerin** [comment]
> Thank you for the feedback!
> 
> **Q1**: May I ask which behavior do you precisely intend?
> 
> 1. When there are some words which can be expanded by `sabbrev`, RET will expand the words, and the user can continue to edit the command line. Finally, when there are no expandable words by `sabbrev` when the user hits RET, the command will be executed.
> 
> ```shell
> $ ble-sabbrev l='ls -l'[RET]
> $ l[RET]
> $ ls -l[RET]
> -rw-r--r-- 1 ply ply 51966 Sep 16 16:08 ble.pp
> -rw-r--r-- 1 ply ply 29823 Sep 16 16:08 blerc
> ...
> ```
> 
> 2. When there are words expandable by `sabbrev`, RET will show the expanded result of the command line and executes the expanded command immediately just like history expansions.

I like this one best.  I think 1 is usually going to be unexpected or else create an unhealthy habit of hitting enter repeatedly, and 3 seems strange because it defeats part of the purpose of using sabbrev instead of alias or function in the first place, and introduces a difference in operation between magic space and Enter.  Also 2 is more consistent with magic space (which both does the expansion and adds the space).

> ```shell
> $ ble-sabbrev l='ls -l'[RET]
> $ l[RET]
> [ble: expand] ls -l
> -rw-r--r-- 1 ply ply 51966 Sep 16 16:08 ble.pp
> -rw-r--r-- 1 ply ply 29823 Sep 16 16:08 blerc
> ...
> ```
> 
> 3. When there are words expandable by `sabbrev`, RETinternally expands the command line and executes the command immediately without showing the expanded result just like alias expansions (I haven't tried but `zsh-abbr` seems to behave like this according to its README).
> 
> ```shell
> $ ble-sabbrev l='ls -l'[RET]
> $ l[RET]
> -rw-r--r-- 1 ply ply 51966 Sep 16 16:08 ble.pp
> -rw-r--r-- 1 ply ply 29823 Sep 16 16:08 blerc
> ...
> ```
> 
> **Q2**: Should it happen also for the words that are not in the position of commands?
> 
> ```shell
> $ ble-sabbrev '\X'='hello world'[RET]
> $ echo \X[RET]
> ```
> 
> **Q3**: Should the cursor position matter for whether the sabbrev expansion happens? I.e., should it happen also when the cursor isn't in the front of the word to be expanded?
> 
> ```shell
> # Example 1
> 
> $ ble-sabbrev l='ls -l'[RET]
> $ l file1[RET]
> 
> # Example 2 (going to the beginning of the command line before pressing RET)
> 
> $ ble-sabbrev l='ls -l'[RET]
> $ [RET]l
> ```
> 
> It seemed like useful in the first thought, and also it's technically possible. But this might not work well because unexpected expansions may happen by the behavior. Also, there is no way to literally run a command that contains words registered in `sabbrev`.

I think it should only fire when the cursor is at the end of the abbreviation when Enter is pressed, both for the reasons you say and also because it's more consistent with how magic space works.

**Reply by akinomyoga** [reply]
Thank you! I'll consider implementing it.

**Reply by akinomyoga** [reply]
I implemented [`bleopt edit_magic_accept`](https://github.com/akinomyoga/ble.sh/wiki/Manual-%C2%A74-Editing#user-content-bleopt-edit_magic_accept) in commits 3e9d89075592a6e6844d4acbf648a742e848f55c and ac84c153a5038814960475b0721f6ac6fad64bf2. You can also see the updated [Recipe R5](https://github.com/akinomyoga/ble.sh/wiki/Recipes#r5--expand-sabbev-via-automatic-magic-space-on-enter) for its usage. Could you check it?

**Reply by akinomyoga** [reply]
xref https://github.com/akinomyoga/ble.sh/issues/406#issuecomment-1962419334

**Comment by bkerin** [comment]
Regarding Q2 I think probably yes for self-consistency and consistency with magic space, but seems like the hardest question as it could produce unexpected behavior and I haven't yet wanted to use it for this (though I'm new to sabbrev and it seems possible to want it)

**Reply by akinomyoga** [reply]
I asked **Q2** just because the OP seemed to assume only the sabbrev expansions of command names. I think we would want to expand arguments too.

**Comment by bkerin** [comment]
Seems to work great.  I have only check sabbrev and sabbrev:verify-syntax though

**Reply by akinomyoga** [reply]
Thank you! It is sufficient for me to check `sabbrev` and `sabbrev:verify-syntax`. I wanted to confirm whether the behavior is the intended one.
